### PR TITLE
Integer overflow detector improvements

### PR DESCRIPTION
**Pull request by feliam**
This change is [Reviewable](https://reviewable.io/reviews/trailofbits/manticore/913)

### CODE REVIEW

File: manticore/ethereum.py [hunk 1]
```diff
@@ -204,42 +193,172 @@ class DetectIntegerOverflow(Detector):
...
         Detects potential overflow and underflow conditions on ADD and SUB instructions.
     '''
+
+    def _save_current_location(self, state, finding, condition):
+        address = state.platform.current_vm.address
+        pc = state.platform.current_vm.pc
+        at_init = state.platform.current_transaction.sort == 'CREATE'
+        location = (address, pc, finding, at_init, condition)
+        hash_id = hashlib.sha1(str(location)).hexdigest()
+        state.context.setdefault('%s.locations' % self.name, {})[hash_id] = location
```
Comment (yan) [inline manticore/ethereum.py]: `%` -> `format`?

File: manticore/ethereum.py [hunk 2]
```diff
@@ -204,42 +193,172 @@ class DetectIntegerOverflow(Detector):
...
+                    if state.can_be_true(loc[-1]):
+                        self.add_finding(state, *loc[:-1])
+            else:
+                for taint in get_taints(what, "IOU_.*"):
+                    loc = self._get_location(state, taint[4:])
+                    if state.can_be_true(loc[-1]):
+                        self.add_finding(state, *loc[:-1])
+
+        if mnemonic in ('SLT', 'SGT', 'SDIV', 'SMOD'):
+            result = taint_with(result, "SIGNED" % id_val)
```
Comment (yan) [inline manticore/ethereum.py]: `"SIGNED" % id_val`?

Comment (yan) [reply manticore/ethereum.py]: Assume it's missing a format string; add it and move it to `.format`? Ditto for below

File: manticore/ethereum.py [hunk 3]
```diff
@@ -1787,24 +1915,31 @@ def finalize(self):
...
+                    finalizer(q.get_nowait())
             except EmptyQueue:
                 pass
 
+        #we need to remove -1 state before forking because it may be in memory
+        if -1 in self._all_state_ids:
+            finalizer(-1)
+
+        q = Queue()
+        map(q.put, self._all_state_ids)
```
Comment (yan) [inline manticore/ethereum.py]: `map` is lazy in py3, this might not execute; replace with explicit loop?

File: manticore/platforms/evm.py [hunk 4]
```diff
@@ -1339,6 +1332,12 @@ def setstate(state, value):
             if self._on_transaction is False:
                 self._publish('will_evm_execute_instruction', current, arguments)
             result = self._handler(*arguments)
+            #print current.semantics,
```
Comment (yan) [inline manticore/platforms/evm.py]: rm commented section?

File: manticore/utils/helpers.py [hunk 5]
```diff
@@ -34,6 +34,40 @@ def istainted(arg, taint=None):
...
+    raise StopIteration
+
+
+def taint_with(arg, taint, value_bits=256, index_bits=256):
+    '''
+    Helper to taint a value, Fixme this should not taint in place.
+    :param arg: a value or Expression
+    :param taint: a regular expression matching a taint value (eg. 'IMPORTANT.*'). If None this functions check for any taint value.
+    '''
+    if not issymbolic(arg):
```
Comment (yan) [inline manticore/utils/helpers.py]: This check looks to be redundant; can just be the check on line below.

Comment (feliam) [reply manticore/utils/helpers.py]: arg is changed if in or long.

Comment (yan) [reply manticore/utils/helpers.py]: `numbers.Integral` is the way to go imho, but we also have `isint()` in the py3 branch which we can port

**Pull request merged**